packages/nimble-components/src/table/virtualization-helper.ts

```diff
@@ -0,0 +1,129 @@
+import { observable } from '@microsoft/fast-element';
+import {
+    Virtualizer,
+    VirtualizerOptions,
+    elementScroll,
+    observeElementOffset,
+    observeElementRect,
+    VirtualItem
+} from '@tanstack/virtual-core';
+import { controlHeight } from '../theme-provider/design-tokens';
+import type { Table } from '.';
+import type { TableRecord } from './types';
+
+/**
+ * Table virtualization helper
+ */
+export class TableVirtualizationHelper<
+    TData extends TableRecord = TableRecord
+> {
+    /**
+     * @internal
+     */
+    public virtualizer?: Virtualizer<HTMLElement, HTMLElement>;
+
+    /**
+     * @internal
+     */
+    @observable
+    public visibleItems: VirtualItem[] = [];
+
+    /**
+     * @internal
+     */
+    @observable
+    public allRowsHeight = 0;
+
+    /**
+     * @internal
+     */
+    @observable
+    public headerContainerMarginRight = 0;
+
+    /**
+     * @internal
+     */
+    @observable
+    public rowContainerYOffset = 0;
+
+    private readonly table: Table<TData>;
+    private readonly viewportResizeObserver: ResizeObserver;
+
+    public constructor(table: Table<TData>) {
+        this.table = table;
+        this.viewportResizeObserver = new ResizeObserver(entries => {
+            const borderBoxSize = entries[0]?.borderBoxSize[0];
+            if (borderBoxSize) {
+                // If we have enough rows that a vertical scrollbar is shown, we need to offset the header widths
+                // by the same margin so the column headers align with the corresponding rendered cells
+                const viewportBoundingWidth = borderBoxSize.inlineSize;
+                this.headerContainerMarginRight = viewportBoundingWidth - this.table.viewport.scrollWidth;
+            }
+        });
+    }
+
+    public handleConnected(): void {
+        this.viewportResizeObserver.observe(this.table.viewport);
+        this.updateVirtualizer();
```

Comment:
updateVirtualizer calling handleVirtualizerChange looks like it's redundant in this code path. handleConnected => updateVirtualizer => _willUpdate calls ResizeObserver observe => onChange is called synchronously => handleVirtualizerChange is called, then it's called again at the end of updateVirtualizer().

Calling handleVirtualizerChange explicitly is needed to handle the case when the # of rows changes though - in that case the element size/ bounds are not changing => ResizeObserver isn't triggered, so we need to tell TanStack Virtualizer about the change (via setting the virtualizer options again).

I'll have to see if I can do some refactoring to get rid of the redundant call in the handleConnected() codepath, but it's harmless (since it's after the ResizeObserver has started observing and the callback has already been called once).

Reply:
Actually it looks like there's some cases where the ResizeObserver callback isn't called immediately after _willUpdate. [These table text column tests](https://github.com/ni/nimble/blob/c0422e918299151bdcbd744fbd0286753c535cf8/packages/nimble-components/src/table-column/text/tests/table-column-text.spec.ts#L56) start failing if updateVirtualizer doesn't have the handleVirtualizerChange() call at the end. So keeping it as-is since it's handling all the various cases right now (setting data before & after connected, etc).